File: 2018/python/day03.py
import re
lines = open("../input/day3.input").readlines()
claims = dict()

for line in lines:
    m = re.search('^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$', line)
    id, left, top, width, height = int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)), int(m.group(5))
    for i in range(top, height + top):
        for j in range(left, width + left):
            if (i,j) in claims:
                claims[(i,j)].append(id)
            else:
                claims[(i,j)] = [id] 

overlapSize = 0
overlaps = set()
for key in claims:
    if len(claims[key]) > 1:
        overlapSize += 1
        for v in claims[key]:
            overlaps.add(v)

print("day3, answer a: ", overlapSize)
print("day3, answer b: ", (set(list(range(1, 1 + len(lines)))) - overlaps).pop())

File: 2018/csharp/day03.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace aoc2018
{
    class Day3
    {
        public static void Solve()
        {
            var lines = File.ReadLines(Path.Combine("..", "input", "day3.input")).ToList();
            var pattern = @"^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$";
            var r = new Regex(pattern);
            var claims = new Dictionary<Point, List<int>>();
            foreach (var line in lines)
            {
                var m = r.Match(line);
                int id = int.Parse(m.Groups[1].Value);
                int left = int.Parse(m.Groups[2].Value);
                int top = int.Parse(m.Groups[3].Value);
                int width = int.Parse(m.Groups[4].Value);
                int height = int.Parse(m.Groups[5].Value);
                for (int i = top; i < height + top; i++)
                {
                    for (int j = left; j < width + left; j++)
                    {
                        Point p = new Point(i, j);
                        if (claims.ContainsKey(p))
                            claims[p].Add(id);
                        else
                            claims[p] = new List<int> { id };
                    }
                }
            }

            int overlapSize = 0;
            HashSet<int> overlaps = new HashSet<int>();
            foreach (var key in claims.Keys)
                if (claims[key].Count > 1)
                {
                    overlapSize++;
                    foreach (var i in claims[key])
                    {
                        overlaps.Add(i);
                    }
                }

            Console.WriteLine("day3, answer a: {0}", overlapSize);
            Console.WriteLine("day3, answer b: {0}", Enumerable.Range(1, lines.Count).Except(overlaps).First());
        }
    }
}
